File: 300.最长递增子序列.py
#
# @lc app=leetcode.cn id=300 lang=python3
#
# [300] 最长递增子序列
#

# @lc code=start
from bisect import bisect_left
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        stack = []
        for num in nums:
            idx = bisect_left(stack, num)
            if idx == len(stack):
                stack.append(num)
            else:
                stack[idx] = num
        return len(stack)

    # def lengthOfLIS(self, nums: List[int]) -> int:
    #     n = len(nums)
    #     dp = [1] * n
    #     for i in range(1, n):
    #         for j in range(i):
    #             if nums[j] < nums[i]:
    #                 dp[i] = max(dp[i], dp[j] + 1)
    #     return max(dp)
# @lc code=end

File: 300.最长递增子序列.cpp
/*
 * @lc app=leetcode.cn id=300 lang=cpp
 *
 * [300] 最长递增子序列
 */
#include<bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp;
        for (int num: nums) {
            int i = lower_bound(dp.begin(), dp.end(), num) - dp.begin();
            if (i >= int(dp.size())) {
                dp.emplace_back(num);
            } else {
                dp[i] = num;
            }
        }
        return dp.size();
    }
};
// @lc code=end

